### PR TITLE
Add exit to adjust rfc server group for parallel serialization

### LINKED ISSUE #6133: change RFC group to other than parallel_generators
Looking for an option to use custom RFC group for parallel processing. Is it possible to change that in settings or so?

https://github.com/abapGit/abapGit/blob/a7c20da7a294d875384f843cbd83cceafdb83fb6/src/objects/core/zcl_abapgit_serialize.clas.abap#L259

### PR DESCRIPTION
**Pull request by fabianlupa**
Fixes #6133

Docs: https://github.com/abapGit/docs.abapgit.org/pull/168

### CODE REVIEW

**Review by larshp — APPROVED** [review]
lgtm

**Comment by mbtools** [comment]
Maybe move the call to method `serialize` and pass in `iv_Package` for added flexibility

**Comment by fabianlupa** [comment]
Great idea. I tried to adjust the code accordingly but after some refactoring to account for this I noticed it's not possible to use different groups in the same (parent) process. `spbt_initialize` only allows one, if called a second time for the same group it fails with `pbt_env_already_initialized`, which the code already accounts for. If it's called a second time for a different group it fails with `cant_init_different_pbt_groups`. This could probably be refactored to not use that function module but I don't think it's worth it. Therefore the package parameter doesn't work as the first exit call would "win" and that group would be used for all other packages as well.

**Comment by mbtools** [comment]
`SPBT_INITIALIZE` can be skipped already if you code function `Z_ABAPGIT_SERIALIZE_PARALLEL` to get the max threads. That's an odd "none-class" exit (I would remove it but that would be a breaking change 🤷 ). 

How about combining `mv_group` and `max_threads` into the new exit? Call `determine_max_threads` only for if the group is `parallel_generators`. 

As a result, you could do the following in the exit:

package | group | max threads
--- | --- | ---
A | GRP_A | 3
B | GRP_B | 5
...

**Comment by fabianlupa** [comment]
Not sure I understand. You mean don't use the group defined in RZ12 if the exit says so and instead retrieve the maximum process count from the exit as well and manage that usage manually?

As I understand it the benefit of RZ12 groups is that if two (parent) processes request work process resources from the same group the maximum process count defined won't be exceeded (i. e. two users opening the abapGit repo in abapGit at the same time).

**Comment by mbtools** [comment]
There are other ways than SPT to determine what system resources to use. The group itself will make sure of certain limits as defined in RZ12. 

Ultimately, it's an exit, so you're responsible for what you implement. 

I was thinking about this:

```abap
  METHOD serialize.
*...
    CLEAR mt_files.

*>>> insert
    li_exit = zcl_abapgit_exit=>get_instance( ).
    li_exit->change_rfc_server_group(
      EXPORTING 
        iv_package          = iv_package
        iv_force_sequential = iv_force_sequential
      CHANGING
        cv_group            = mv_group
        cv_max_threads      = lv_max ).

    IF mv_group = 'parallel_generators'.
      lv_max = determine_max_threads( iv_force_sequential ).
    ELSEIF lv_max < 1.
      lv_max = 1.
    ENDIF.
*<<< insert

    mv_free = lv_max.
    mi_log = ii_log.

```

**Comment by fabianlupa** [comment]
Ah I see. I'll have another go at it.

**Comment by larshp** [comment]
@fabianlupa convert to draft?

**Comment by fabianlupa** [comment]
Draft!

**Comment by fabianlupa** [comment]
Let's step back a bit to the drawing board. Use cases for the new exit:

1. Use a different RZ11 group than `parallel_generators` but keep the "process pool management" the same, i. e. "maximum threads"
2. Do not use an RZ11 group
3. Provide your own logic for the maximum number of parallel processes (including by package) for the determined parallelization technique

1 would actually be the use case for me as for some reason the fully activated appliances are shipped with a marginally differently named group for us... 3 could be used to override the default maximum number of processes including by package regardless of the group used or none (to make your favorite packages faster ;) ). 2 somewhat requires 3 to be implemented.

Two RZ11 groups cannot be used in the same parent process according to the docs even if you don't call the initialization function module yourself. Therefore no package parameter makes sense to override the determination.

Therefore I propose two exits:

```abap
  METHODS change_rfc_server_group
    CHANGING
      cv_group TYPE rzlli_apcl.
```

In addition to the current implementation in this PR it would also be allowed to return an initial value in `change_rfc_server_group`. Then `Z_ABAPGIT_SERIALIZE_PARALLEL` would be called without a group but still in parallel.

```abap
  METHODS change_max_threads
    IMPORTING
      iv_package TYPE devclass
    CHANGING
      cv_max_threads TYPE i.
```

`change_max_threads` allows for adjusting the determined maximum process count to be used. This would be required when running without an RZ11 group as the default implementation would determine 1 and run sequentially.

Only one exit would have a very confusing signature.

By the way I dislike the term "threads" being used here, ABAP does not allow multi-threading, only multiple "connected" processes and I therefore propose to change it, for example `change_max_parallel_processes`.

**Comment by mbtools** [comment]
Makes perfect sense. Go for it!

Parallel_processes is better than threads, indeed

**Comment by fabianlupa** [comment]
Sorry for the delay. I'll get back to it this weekend.

**Comment by fabianlupa** [comment]
I adjusted the PR based on the comment above. I think this makes the most sense now.

Note these lines:
https://github.com/abapGit/abapGit/blob/c6ace51b1dcba90e83c4ea2608a48ded163b4f3e/src/objects/core/zcl_abapgit_serialize.clas.abap#L609-L612
No longer work that well. If `change_max_parallel_processes` returns more processes than there are available `mv_free` no longer reflects the actual free count accurately (it didn't before either in all cases), therefore the `WAIT ... UP TO 1 SECONDS` will just be skipped and tried again immediately. That might be fair though as the one implementing the exit chose to know it better and may just determine the actually free processes accurately in the exit implementation.

**Comment by fabianlupa** [comment]
Looking at the unit test: Do we not have a way to disable exit calls / mock an exit implementation currently?

**Comment by mbtools** [comment]
Looks good. Small ask: please put the exit interface methods in alphabetical order. Just makes it a bit easier to find

I don't think we have any UTs for the exits. It should just work well without exit implementations 😀

File: src/objects/core/zcl_abapgit_serialize.clas.abap [hunk 1]
```diff
@@ -303,63 +299,81 @@ CLASS zcl_abapgit_serialize IMPLEMENTATION.
...
 
-    IF gv_max_threads >= 1.
-* SPBT_INITIALIZE gives error PBT_ENV_ALREADY_INITIALIZED if called
-* multiple times in same session
-      rv_threads = gv_max_threads.
-      RETURN.
+    IF gv_max_processes IS INITIAL.
+      lo_settings = zcl_abapgit_persist_factory=>get_settings( )->read( ).
+
+      IF zcl_abapgit_factory=>get_environment( )->is_merged( ) = abap_true
```
Comment (mbtools) [inline src/objects/core/zcl_abapgit_serialize.clas.abap:314]: the check here means that we can skip the "function exists" check below (if the function really does not exist in the dev version, then causing a dump is good).

Comment (fabianlupa) [reply src/objects/core/zcl_abapgit_serialize.clas.abap:314]: I agree and removed the call. The transpiler setup doesn't have the function module either though it seems ;)

**Comment by larshp** [comment]
why is this not necessary anymore?

```abap
      CALL FUNCTION 'FUNCTION_EXISTS'
        EXPORTING
          funcname           = 'Z_ABAPGIT_SERIALIZE_PARALLEL'
```

**Comment by fabianlupa** [comment]
> why is this not necessary anymore?
> 
> ```abap
>       CALL FUNCTION 'FUNCTION_EXISTS'
>         EXPORTING
>           funcname           = 'Z_ABAPGIT_SERIALIZE_PARALLEL'
> ```

I think Marc's idea was that is_merged = false already covers that branch, see the review comment above.

**Comment by mbtools** [comment]
Tested and working well! Thank you 👍

**Comment by abap-weasel** [comment]
Thanks @fabianlupa - already using new exit method 👍

**Pull request merged**